repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeFilterChain.scala

```diff
@@ -0,0 +1,135 @@
+import java.util.concurrent.TimeUnit
+
+import com.codahale.metrics.MetricRegistry
+import com.fasterxml.jackson.annotation.{JsonAutoDetect, PropertyAccessor}
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.typesafe.scalalogging.slf4j.StrictLogging
+import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
+import javax.servlet.{Filter, FilterChain, ServletRequest, ServletResponse}
+import org.openrepose.commons.utils.io.{BufferedServletInputStream, RawInputStreamReader}
+import org.openrepose.commons.utils.servlet.http.ResponseMode.{PASSTHROUGH, READONLY}
+import org.openrepose.commons.utils.servlet.http.{HttpServletRequestWrapper, HttpServletResponseWrapper}
+import org.openrepose.powerfilter.ReposeFilterChain._
+import org.openrepose.powerfilter.intrafilterlogging.{RequestLog, ResponseLog}
+import org.slf4j.{Logger, LoggerFactory}
+
+class ReposeFilterChain(val filterChain: List[FilterContext], originalChain: FilterChain, bypassUrlRegex: Option[String], metricsRegistry: MetricRegistry)
+  extends FilterChain
+    with StrictLogging {
+
+  override def doFilter(inboundRequest: ServletRequest, inboundResponse: ServletResponse): Unit = {
+    val request = inboundRequest.asInstanceOf[HttpServletRequest]
+    val response = inboundResponse.asInstanceOf[HttpServletResponse]
+    bypassUrlRegex.map(_.r.pattern.matcher(request.getRequestURI).matches()) match {
+      case Some(true) =>
+        logger.debug("Bypass url hit")
+        runNext(List.empty, request, response)
+      case _ =>
+        runNext(filterChain, request, response)
+    }
+  }
+
+  def runNext(chain: List[FilterContext], request: HttpServletRequest, response: HttpServletResponse): Unit = {
+    chain match {
+      case Nil =>
+        doIntrafilterLogging(request, response, "origin", (intraRequest, intraResponse) => {
+          doMetrics(intraRequest, intraResponse, "origin", (metricsRequest, metricsResponse) => {
+            logger.debug("End of the filter chain reached")
+            originalChain.doFilter(metricsRequest, metricsResponse)
+          })
+        })
+      case head::tail =>
+        if (head.shouldRun(request)) {
+          doIntrafilterLogging(request, response, head.filterName, (intraRequest, intraResponse) => {
+            doMetrics(intraRequest, intraResponse, head.filterName, (metricsRequest, metricsResponse) => {
+              logger.debug("Entering filter: {}", head.filterName)
+              head.filter.doFilter(metricsRequest, metricsResponse, new ReposeFilterChain(tail, originalChain, None, metricsRegistry))
+            })
+          })
+        } else {
+          logger.debug("Skipping filter: {}", head.filterName)
+          runNext(tail, request, response)
+        }
+    }
+  }
+
+  def doIntrafilterLogging(request: HttpServletRequest, response: HttpServletResponse, filter: String, requestProcess: (HttpServletRequest, HttpServletResponse) => Unit): Unit = {
```

Comment:
I decided to do this and metrics like this so that what we do for logging, metrics and routing are a little more clear. since they are all isolated.

Reply:
A PR with these changes a bit more isolated:
https://github.com/rackerlabs/repose/pull/1943/files